src/main/scala/org/broadinstitute/dsde/firecloud/service/ExportEntitiesByTypeActor.scala

```diff
@@ -1,79 +1,260 @@
+    // Result of this will be a tuple of Future[IOResult] that represents the success or failure of
+    // streaming content to the file sinks.
+    val fileStreamIOResults: (Future[IOResult], Future[IOResult]) = {
+      RunnableGraph.fromGraph(GraphDSL.create(entitySink, membershipSink)((_, _)) { implicit builder =>
+        (eSink, mSink) =>
+          import GraphDSL.Implicits._
+
+          // Sources
+          val entitySource: Outlet[Seq[Entity]] = builder.add(Source.fromFuture(entityBatches)).out
+          val entityHeaderSource: Outlet[ByteString] = builder.add(Source.single(ByteString(entityHeaders.mkString("\t") + "\n"))).out
+          val membershipHeaderSource: Outlet[ByteString] = builder.add(Source.single(ByteString(membershipHeaders.mkString("\t") + "\n"))).out
+
+          // Flows
+          val splitter: UniformFanOutShape[Seq[Entity], Seq[Entity]] = builder.add(Broadcast[Seq[Entity]](2))
+          val entityFlow: FlowShape[Seq[Entity], ByteString] = builder.add(Flow[Seq[Entity]].map { entities =>
+            val rows = TSVFormatter.makeEntityRows(entityType, entities, entityHeaders)
+            ByteString(rows.map { _.mkString("\t")}.mkString("\n") + "\n")
+          })
+          val membershipFlow: FlowShape[Seq[Entity], ByteString] = builder.add(Flow[Seq[Entity]].map { entities =>
+            val rows = TSVFormatter.makeMembershipRows(entityType, entities)
+            ByteString(rows.map { _.mkString("\t")}.mkString("\n") + "\n")
+          })
+          val eConcat: UniformFanInShape[ByteString, ByteString] = builder.add(Concat[ByteString]())
+          val mConcat: UniformFanInShape[ByteString, ByteString] = builder.add(Concat[ByteString]())
+
+          // Graph
+          entityHeaderSource                         ~> eConcat
+          entitySource ~> splitter ~> entityFlow     ~> eConcat ~> eSink
+          membershipHeaderSource                     ~> mConcat
+                          splitter ~> membershipFlow ~> mConcat ~> mSink
+          ClosedShape
+      }).run()
+    }
+
+    // Check that each file is completed
+    val fileStreamResult = for {
+      eResult <- fileStreamIOResults._1
+      mResult <- fileStreamIOResults._2
+    } yield eResult.wasSuccessful && mResult.wasSuccessful
+
+    // And then map those files to a ZIP.
+    fileStreamResult map { s =>
+      if (s) {
+        val zipFile = writeFilesToZip(entityType, tempEntityFile, tempMembershipFile)
+        // The output to the user
+        lazy val streamingActorRef = actorRefFactory.actorOf(Props(new StreamingActor(ctx, ContentTypes.`application/octet-stream`, entityType + ".zip")))
+        zipFile map { f =>
+          streamingActorRef ! FirstChunk(HttpData.apply(f.byteArray), 0)
+        }
+      } else {
+        throw new FireCloudExceptionWithErrorReport(ErrorReport(s"Unable to collect entity metadata for $workspaceNamespace:$workspaceName:$entityType"))
+      }
+    }
+
+    Future(Done)
```

Comment:
... or just have this method and `streamSingularType` return a `Future[Unit]`, not sure we need `Future[Done]`.